### DISCUSSION TITLE
More configurability for state cookies and better defaults for unathenticated actions with respects to state cookies

### DISCUSSION THREAD

**Discussion by studersi** [OpenIDC/mod_auth_openidc#714] [Ideas]
The state cookie does not expire until the end of the session. However, generally the state cookies are only needed for a few seconds, until the login workflow has completed. There are cases where state cookies are accumulated in the current session. Once the limit of 7 is reached, the website becomes unavailable for the user-agent until the session is terminated, because requests are answered with 503. This can be mitigated to an extent with `OIDCStateMaxNumberOfCookies 7 true`. However, this can create loops of very quick infinite retries.

This behavior can be quite jarring to the user experience.

In our estimation this is mostly due to background requests that are incorrectly being redirected to the login page instead of returning 401. This case often occurs because applications do not declare that the background requests are XHR requests and add `*/*` to the `Accept` header.

A default behavior along the lines of the following configuration, taking the `Sec-Fetch-Mode` header into account, would likely mitigate a lot of problem cases automatically because the `Sec-Fetch-Mode` header is set automatically by the browser.
`OIDCUnAuthAction 401 "%{HTTP:Sec-Fetch-Mode} != 'navigate' || %{HTTP:X-Requested-With} == 'XMLHttpRequest' || ( ( %{HTTP_ACCEPT} !~ m#text/html# ) && ( %{HTTP_ACCEPT} !~ m#application/xhtml\+xml# ) && ( %{HTTP_ACCEPT} !~ m#\*/\*# ) )"
`
Currently, this has to be configured explicitly.

Further, it would be great to have additional options to configure the behavior when the limit of state cookies is exceeded. For example it would be helpful to configure the status code and to configure the expiration of the state cookies in order to lessen the impact of the problems described above.

**Comment by zandbelt** [comment]
I agree with changing the default auto-detect algorithm and the Sec-Fetch-Mode header suggestion was added in https://github.com/zmartzone/mod_auth_openidc/commit/ee891b1665be34b624df11a103a910d74869256c

**Reply by studersi** [reply]
Thanks for adding this to the default expression.

I just had a thought, however. Maybe we should only consider the Sef-Fetch-Mode Header if it is present. While user-agents that do not send this header are likely not browsers, there still are browsers that do not implement this header: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Sec-Fetch-Mode#browser_compatibility.

What do you think?

**Reply by zandbelt** [reply]
yes, that's how it is implemented: if it is set and != navigate it will be considered an XML HTTP Request, otherwise the processing will continue as before

**Reply by studersi** [reply]
Ah, I see. Wouldn't the corresponding `OIDCUnAuthAction` be something like the following instead?
```
OIDCUnAuthAction 401 "%{HTTP:X-Requested-With} == 'XMLHttpRequest' || ( -n %{HTTP:Sec-Fetch-Mode} && %{HTTP:Sec-Fetch-Mode} != 'navigate' ) || ( ( %{HTTP_ACCEPT} !~ m#text/html# ) && ( %{HTTP_ACCEPT} !~ m#application/xhtml\+xml# ) && ( %{HTTP_ACCEPT} !~ m#\*/\*# ) )"
```

As far as I can tell the example in the documentation which is said to correspond to the default behaviour would respond with `401` instead of `auth` if the `Sec-Fetch-Mode` header is missing.

**Reply by zandbelt** [reply]
ok, that's right, corrected in https://github.com/zmartzone/mod_auth_openidc/commit/f4ea4e73e20660d26665ea365fbcd78a4242272f

**Reply by studersi** [reply]
Perfect, thanks!

**Reply by zandbelt** [reply]
I'm considering to add a similar check for `Sec-Fetch-Dest` != `document`, I'm interested to learn if you see objections to that.

**Reply by studersi** [reply]
I would be hesitant to add it. Not because I think there is a problem with it but simply because it is redundant if there is a check for `Sec-Fetch-Mode != 'navigate'`.

`Sec-Fetch-Dest: document`, as opposed to other HTML destinations, requires for the request to be initiated by the user. I assume that this is why you suggest it be added to the check. The fact that the request is initiated by the user is redundant because this information is already covered by the `Sec-Fetch-Mode` check.

I am not aware of a situation where the browser would set the `Sec-Fetch-Dest` header but not the `Sec-Fetch-Mode` header. However, the inverse is the case for certain browsers or browser version as `Sec-Fetch-Dest` was introduced later and thus not supported by all browsers or versions that support `Sec-Fetch-Mode`.

**Reply by zandbelt** [reply]
thanks, I was assuming that it would catch iframe based requests (which should not - at least not by default - be used for authentication round trips) but more over all types of resources loaded from top level HTML pages like images and stylesheets

**Reply by zandbelt** [reply]
ah, but reading the docs more closely it seems indeed covered by `no-cors`...

**Reply by studersi** [reply]
I was not aware that for iframes, the `Sec-Fetch-Mode` is also `navigate`. In that case I agree that it would make sense to check also if `Sec-Fetch-Dest`  is not `document` if the header is set.

**Reply by studersi** [reply]
In fact, I just checked here: https://www.w3schools.com/html/html_iframe.asp. For loading the page, there is a request that has the following combination of headers:
```
Sec-Fetch-Dest: iframe
Sec-Fetch-Mode: navigate
```

**Reply by zandbelt** [reply]
thanks for confirming; since this is a potentially breaking change it will be part of the upcoming 2.4.10, see: https://github.com/zmartzone/mod_auth_openidc/commit/143b63b8b25b2625730fa5db1941aee5613fcd70

**Reply by studersi** [reply]
Ok, thanks.
